File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/Environment.java
package uk.nhs.nhsx.core;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import uk.nhs.nhsx.core.exceptions.Defect;

import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.util.function.Function.identity;

public class Environment {

    public static final EnvironmentKey<String> WORKSPACE = EnvironmentKey.string("WORKSPACE");

    private static final String UNKNOWN = "unknown";
    public final EnvironmentName name;

    public final EnvironmentType type;
    public final Access access;
    private final Logger logger = LogManager.getLogger(Environment.class);

    private Environment(EnvironmentName name, EnvironmentType type, Access access) {
        this.name = name;
        this.type = type;
        this.access = access;
    }

    public static Environment unknown() {
        return fromName(UNKNOWN, Access.SYSTEM);
    }

    public static Environment fromSystem() {
        return fromEnvironment(Access.SYSTEM);
    }

    public static Environment fromEnvironment(Environment.Access access) {
        return fromName(access.required(WORKSPACE), access);
    }

    public static Environment fromName(String name, Environment.Access access) {
        return new Environment(
            EnvironmentName.of(name), environmentTypeFrom(name), access
        );
    }

    private static EnvironmentType environmentTypeFrom(String name) {
        if (name == null || name.equals(UNKNOWN) || name.startsWith("te-prod")) {
            return EnvironmentType.Production;
        }
        return EnvironmentType.NonProduction;
    }

    public <T> Optional<T> whenNonProduction(String functionality, Function<Environment, T> consumer) {
        if (type == EnvironmentType.NonProduction) {
            logger.error("ENABLING TEST FUNCTIONALITY:" + functionality);
            return Optional.ofNullable(consumer.apply(this));
        }
        return Optional.empty();
    }

    public enum EnvironmentType {NonProduction, Production}

    public interface Access {
        Access SYSTEM = new Access() {
            @Override
            public <T> Optional<T> optional(EnvironmentKey<T> key) {
                return Optional.ofNullable(System.getenv(key.name)).map(key.conversion);
            }
        };

        Function<Map<String, String>, Access> TEST = m -> new Access() {
            @Override
            public <T> Optional<T> optional(EnvironmentKey<T> key) {
                return Optional.ofNullable(m.get(key.name)).map(key.conversion);
            }
        };

        <T> Optional<T> optional(EnvironmentKey<T> key);

        default <T> T required(EnvironmentKey<T> key) {
            return optional(key).orElseThrow(() -> new IllegalArgumentException(String.format("Environment variable %s is missing", key.name)));
        }
    }

    public static class EnvironmentName extends ValueType<EnvironmentName> {

        protected EnvironmentName(String value) {
            super(value);
        }

        public static EnvironmentName of(String value) {
            return new EnvironmentName(value);
        }
    }

    public static class EnvironmentKey<T> {
        public final String name;
        public final Function<String, T> conversion;

        public EnvironmentKey(String name, Function<String, T> conversion) {
            this.name = name;
            this.conversion = conversion;
        }

        public static <T> EnvironmentKey<T> define(String name, Function<String, T> conversion) {
            return new EnvironmentKey<>(name, conversion);
        }

        public static EnvironmentKey<String> string(String name) {
            return define(name, identity());
        }

        public static EnvironmentKey<Boolean> bool(String name) {
            return define(name, Boolean::parseBoolean);
        }

        public static EnvironmentKey<Integer> integer(String name) {
            return define(name, Integer::parseInt);
        }

        public static <T> EnvironmentKey<List<T>> list(String name, Function<String, T> conversion) {
            return define(name, s -> Arrays.stream(s.split(",")).filter(it -> !it.isBlank()).map(conversion).collect(Collectors.toList()));
        }

        public static EnvironmentKey<List<String>> strings(String name) {
            return list(name, identity());
        }

        public static <T extends ValueType<T>> EnvironmentKey<T> value(String name, Class<T> clazz) {
            return define(name, valueTypeConverterFor(clazz));
        }

        private static <T extends ValueType<T>> Function<String, T> valueTypeConverterFor(Class<T> clazz) {
            return s -> {
                try {
                    return clazz.cast(clazz.getMethod("of", String.class).invoke(null, s));
                } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
                    throw new Defect("ValueType missing expected method String of(String)", e);
                }
            };
        }
    }
}

File: src/aws/lambdas/incremental_distribution/infra/src/main/kotlin/uk/nhs/nhsx/core/Environment.kt
package uk.nhs.nhsx.core

import dev.forkhandles.values.Value
import dev.forkhandles.values.ValueFactory
import uk.nhs.nhsx.core.EnvironmentName.Companion.of
import java.lang.Boolean.parseBoolean
import java.time.Duration
import java.time.LocalDate
import java.util.Optional
import java.util.function.Function
import java.util.function.Supplier

class Environment private constructor(val name: EnvironmentName, val type: EnvironmentType, val access: Access) {
    interface Access {
        fun <T> optional(key: EnvironmentKey<T>): Optional<T>

        fun <T> required(key: EnvironmentKey<T>): T = optional(key)
            .orElseThrow { IllegalArgumentException("Environment variable ${key.name} is missing") }

        fun <T> defaulted(key: EnvironmentKey<T>, value: Supplier<T>): T = optional(key).orElseGet(value)

        companion object {
            val SYSTEM: Access = object : Access {
                override fun <T> optional(key: EnvironmentKey<T>) =
                    Optional.ofNullable(System.getenv(key.name)).map(key.conversion)
            }
            val TEST = Function<Map<String, String>, Access> { m: Map<String, String> ->
                object : Access {
                    override fun <T> optional(key: EnvironmentKey<T>) =
                        Optional.ofNullable(m[key.name]).map(key.conversion)
                }
            }
        }
    }

    class EnvironmentKey<T>(val name: String, val conversion: (String) -> T) {
        companion object {
            private fun <T> define(name: String, conversion: (String) -> T): EnvironmentKey<T> =
                EnvironmentKey(name, conversion)

            fun string(name: String) = define(name) { it }

            fun bool(name: String) = define(name, ::parseBoolean)

            fun duration(name: String) = define(name, Duration::parse)

            fun integer(name: String) = define(name, String::toInt)

            fun long(name: String) = define(name, String::toLong)

            fun strings(name: String) = define(name) { it.split(",").map(String::trim).filter(String::isNotBlank) }

            fun localDate(name: String) = define(name, LocalDate::parse)

            fun <X : Any, T : Value<X>> value(name: String, fn: ValueFactory<T, X>) = define(name, fn::parse)
        }
    }

    companion object {
        val WORKSPACE = EnvironmentKey.string("WORKSPACE")

        private const val UNKNOWN = "unknown"
        fun unknown(): Environment = fromName(UNKNOWN, Access.SYSTEM)

        fun fromSystem(): Environment = fromEnvironment(Access.SYSTEM)

        fun fromEnvironment(access: Access): Environment = fromName(access.required(WORKSPACE), access)

        fun fromName(name: String, access: Access): Environment =
            Environment(of(name), environmentTypeFrom(name), access)

        private fun environmentTypeFrom(name: String) =
            when {
                name == UNKNOWN || name.startsWith("te-prod") -> EnvironmentType.Production
                else -> EnvironmentType.NonProduction
            }
    }
}
